File: agentic_os/mission/learners.py
"""P9b — three learners kept SEPARATE, behind a promotion ladder.

"Learning" here is three different problems with distinct feedback signals, safety risks and
rollout strategies — they must not share a model:

  * Routing learning        — which operator/provider executes a capability best.
  * Model-selection learning — which model tier fits a reasoning step (cost·latency·confidence·success).
  * Plan-policy learning     — which execution graph / ordering works best per mission type.

None mutates production behaviour directly. Every learner follows the ladder

    Observe → Recommend → Shadow-evaluate → Policy-constrained promotion

so unrestricted online mutation can't undermine the determinism P7 established. The *only* path that
changes production routing is `LearningStack.promote()`, which a policy (or a P10 governance action)
must call explicitly.
"""
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field

from .types import new_id


@dataclass
class Recommendation:
    kind: str                  # routing | model_selection | plan_policy
    subject: str               # outcome / reasoning-step / mission-kind
    current: str               # the option production uses now
    recommended: str           # the option the evidence favours
    support: int               # observations backing the recommendation
    lift: float                # measured improvement over `current`
    status: str = "recommended"  # recommended → shadow → promoted
    id: str = field(default_factory=lambda: new_id("rec"))


class _StatTable:
    """Success rate + support per (subject, option)."""
    def __init__(self):
        self._s: dict[str, dict[str, list[int]]] = defaultdict(lambda: defaultdict(list))

    def record(self, subject: str, option: str, ok: bool) -> None:
        self._s[subject][option].append(1 if ok else 0)

    def rate(self, subject: str, option: str) -> float | None:
        h = self._s[subject].get(option)
        return round(sum(h) / len(h), 3) if h else None

    def options(self, subject: str) -> dict[str, list[int]]:
        return dict(self._s[subject])

    def subjects(self) -> list[str]:
        return list(self._s)


class Recommender:
    """Observe outcomes; recommend a better option once enough evidence shows a real lift."""
    kind = "base"

    def __init__(self, min_support: int = 5, min_lift: float = 0.1):
        self.tbl = _StatTable()
        self.min_support = min_support
        self.min_lift = min_lift

    def observe(self, subject: str, option: str, ok: bool) -> None:
        self.tbl.record(subject, option, ok)

    def recommend(self) -> list[Recommendation]:
        recs: list[Recommendation] = []
        for subj in self.tbl.subjects():
            opts = self.tbl.options(subj)
            if len(opts) < 2:
                continue
            current = max(opts, key=lambda o: len(opts[o]))   # production = the most-used option
            cur_rate = self.tbl.rate(subj, current) or 0.0
            best = None
            for opt, hist in opts.items():
                if opt == current or len(hist) < self.min_support:
                    continue
                lift = round((self.tbl.rate(subj, opt) or 0.0) - cur_rate, 3)
                if lift >= self.min_lift and (best is None or lift > best.lift):
                    best = Recommendation(self.kind, subj, current, opt, len(hist), lift)
            if best is not None:
                recs.append(best)
        return recs


class RoutingLearner(Recommender):
    kind = "routing"            # subject = outcome, option = provider/operator capability


class ModelSelectionLearner(Recommender):
    kind = "model_selection"    # subject = reasoning-step signature, option = model tier


class PlanPolicyLearner(Recommender):
    kind = "plan_policy"        # subject = mission kind, option = plan signature


class LearningStack:
    """Holds the three learners and the promotion gate. Recommendations are shadow-only until a
    policy promotes them — the one place production routing can change."""

    def __init__(self, *, allow_auto_promote: bool = False):
        self.routing = RoutingLearner()
        self.models = ModelSelectionLearner()
        self.plans = PlanPolicyLearner()
        self.allow_auto_promote = allow_auto_promote
        self._promoted: dict[tuple[str, str], str] = {}

    def _learners(self):
        return (self.routing, self.models, self.plans)

    def recommendations(self) -> list[Recommendation]:
        out: list[Recommendation] = []
        for lr in self._learners():
            for r in lr.recommend():
                key = (r.kind, r.subject)
                if self._promoted.get(key) == r.recommended:
                    r.status = "promoted"
                elif self.allow_auto_promote:
                    r.status = "shadow"       # eligible; a promote() call still gates the mutation
                out.append(r)
        return out

    def promote(self, rec: Recommendation, *, actor: str = "policy") -> Recommendation:
        """Policy-constrained promotion — the ONLY path that changes production routing."""
        self._promoted[(rec.kind, rec.subject)] = rec.recommended
        rec.status = "promoted"
        return rec

    def promoted_choice(self, kind: str, subject: str) -> str | None:
        """The promoted option for a (kind, subject), if any — how the compiler/router reads a promotion."""
        return self._promoted.get((kind, subject))

File: go/mission/learners.go
package mission

// P9b — three learners kept SEPARATE, behind a promotion ladder. Routing / model-selection /
// plan-policy have distinct feedback signals; none mutates production until a policy promotes it
// (Observe → Recommend → Shadow → Policy-constrained promotion). Go port of learners.py.

// Recommendation is a proposed option change with the evidence backing it.
type Recommendation struct {
	Kind        string  `json:"kind"`
	Subject     string  `json:"subject"`
	Current     string  `json:"current"`
	Recommended string  `json:"recommended"`
	Support     int     `json:"support"`
	Lift        float64 `json:"lift"`
	Status      string  `json:"status"`
	ID          string  `json:"id"`
}

func newRecommendation(kind, subject, current, recommended string, support int, lift float64) Recommendation {
	return Recommendation{Kind: kind, Subject: subject, Current: current, Recommended: recommended,
		Support: support, Lift: lift, Status: "recommended", ID: newID("rec")}
}

// statTable tracks success rate + support per (subject, option), preserving first-seen order.
type statTable struct {
	s         map[string]map[string][]int
	subjOrder []string
	optOrder  map[string][]string
}

func newStatTable() *statTable {
	return &statTable{s: map[string]map[string][]int{}, optOrder: map[string][]string{}}
}

func (t *statTable) record(subject, option string, ok bool) {
	if _, e := t.s[subject]; !e {
		t.s[subject] = map[string][]int{}
		t.subjOrder = append(t.subjOrder, subject)
	}
	if _, e := t.s[subject][option]; !e {
		t.optOrder[subject] = append(t.optOrder[subject], option)
	}
	t.s[subject][option] = append(t.s[subject][option], b2i(ok))
}

func (t *statTable) hist(subject, option string) []int { return t.s[subject][option] }

// Recommender observes outcomes and recommends a better option once evidence shows a real lift.
type Recommender struct {
	kind       string
	tbl        *statTable
	MinSupport int
	MinLift    float64
}

func newRecommender(kind string) *Recommender {
	return &Recommender{kind: kind, tbl: newStatTable(), MinSupport: 5, MinLift: 0.1}
}

func (r *Recommender) Observe(subject, option string, ok bool) { r.tbl.record(subject, option, ok) }

func (r *Recommender) Recommend() []Recommendation {
	var recs []Recommendation
	for _, subj := range r.tbl.subjOrder {
		opts := r.tbl.optOrder[subj]
		if len(opts) < 2 {
			continue
		}
		current := opts[0] // production = the most-used option (first-seen wins ties)
		for _, o := range opts {
			if len(r.tbl.hist(subj, o)) > len(r.tbl.hist(subj, current)) {
				current = o
			}
		}
		curRate := avgInts(r.tbl.hist(subj, current))
		var best *Recommendation
		for _, opt := range opts {
			hist := r.tbl.hist(subj, opt)
			if opt == current || len(hist) < r.MinSupport {
				continue
			}
			lift := roundN(avgInts(hist)-curRate, 3)
			if lift >= r.MinLift && (best == nil || lift > best.Lift) {
				rec := newRecommendation(r.kind, subj, current, opt, len(hist), lift)
				best = &rec
			}
		}
		if best != nil {
			recs = append(recs, *best)
		}
	}
	return recs
}

// LearningStack holds the three learners and the promotion gate. Recommendations are shadow-only
// until a policy promotes them — the one place production routing can change.
type LearningStack struct {
	Routing          *Recommender
	Models           *Recommender
	Plans            *Recommender
	AllowAutoPromote bool
	promoted         map[[2]string]string
}

func NewLearningStack() *LearningStack {
	return &LearningStack{
		Routing: newRecommender("routing"), Models: newRecommender("model_selection"),
		Plans: newRecommender("plan_policy"), promoted: map[[2]string]string{},
	}
}

func (s *LearningStack) Recommendations() []Recommendation {
	var out []Recommendation
	for _, lr := range []*Recommender{s.Routing, s.Models, s.Plans} {
		for _, r := range lr.Recommend() {
			key := [2]string{r.Kind, r.Subject}
			if s.promoted[key] == r.Recommended {
				r.Status = "promoted"
			} else if s.AllowAutoPromote {
				r.Status = "shadow"
			}
			out = append(out, r)
		}
	}
	return out
}

// Promote is the ONLY path that changes production routing.
func (s *LearningStack) Promote(rec *Recommendation, actor string) *Recommendation {
	s.promoted[[2]string{rec.Kind, rec.Subject}] = rec.Recommended
	rec.Status = "promoted"
	return rec
}

func (s *LearningStack) PromotedChoice(kind, subject string) (string, bool) {
	v, ok := s.promoted[[2]string{kind, subject}]
	return v, ok
}
